validation/validate_map.py

```diff
@@ -0,0 +1,202 @@
+      syn: Synapse object
+      config: configuration parameters
+    Returns:
+        dataframe with metadata on any empty codes.
+  """
+  empty = df.loc[pd.isna(df['code'])]['code']
+  return(list(empty))
+  
+def _check_code_name_absent(df: pd.DataFrame, syn: Synapse, config: dict) -> list:
+  """Check for any code that is not code name that
+    does not appear in its associated data file.  
+    Args:
+        df: dataframe representing map
+        syn: Synapse object
+        config: configuration parameters
+    Returns:
+        dataframe with metadata on any missing codes.
+  """
+  absent = []
+  for dataset in config['dataset']:
+    data = pd.read_csv(syn.get(config['dataset'][dataset]['id'])['path'], low_memory=False)
+    code_data = data.columns
+    
+    # get codes associated with the dataset and of types derived or curated
+    code_map = list(df.loc[((df['dataset'] == dataset) & 
+                        ((df['data_type'].str.lower() == 'derived') | (df['data_type'].str.lower() == 'curated')))]['code'])
+                        
+    # do not check wildcard code names or NA code names
+    code_remove = []
+    for code in code_map:
+      if bool(re.match(r'^.+[*]$', str(code).strip())):
+        code_remove.append(code)
+      elif pd.isna(code):
+        code_remove.append(code)
+    for code in code_remove:
+      code_map.remove(code)
+    
+    absent.extend(list(set(code_map) - set(code_data)))
+  return(absent)
+
+
+def _format_result(codes: list, config: dict, check_no: int):
+  """Format output for interpretable log file.  
+    Args:
+        df: dataframe representing map
+        config: configuration parameters
+        check_no: check number for which to format results
+    Returns:
+        dataframe with additional metadata on any errors.
+  """
+  formatted = pd.DataFrame()
+  formatted['code'] = codes
+  formatted['check_no'] = str(check_no)
+  formatted['description'] = config['check'][check_no]['description']
+  formatted['action'] = config['check'][check_no]['request']
+  return(formatted)
+  
+def validate_map(synapse_id: str, syn: Synapse, config: dict, version: int) -> pd.DataFrame:
+  """Run all implemented checks on mapping file.  
+    Args:
+        synapse_id: Synapse ID of mapping file
+        syn: Synapse object
+        config: configuration parameters
+        version: Version number of Synapse ID
+    Returns:
+        dataframe with additional metadata on any errors.
+  """
+
+  errors = pd.DataFrame()
+  df = pd.DataFrame()
+  if (version == 'None'):
+    df = pd.read_csv(syn.get(synapse_id)['path'])
+  else:
+    df = pd.read_csv(syn.get(synapse_id, version=version)['path'])
+  
+  for check_no in config['check']:
+    
+    logging.info(f'Check {check_no}...')
+    
+    if (config['check'][check_no]['implemented'] and not config['check'][check_no]['deprecated']): 
+      function_name = config['check'][check_no]['function']
+      result = eval(function_name + "(df, syn, config)")
```

Comment:
I think it's pretty risky to use `eval` in general because it does a system call.  I'm trying to think if there is another way to do this.  If there was a Validation Class, could we access the function another way?

Reply:
Ok.  Each check object needs associated metadata (check number, description, request) and then an associated function.  Having the metadata stored in the config file makes sense and makes it easier to edit and organize without having to hard code this information.  There could be a Validation class with standard metadata and a validate function but there still has to be a way of using the string representation of the function name to link to the actual function.  Is there a way to do that without eval?

Reply:
Hmm.. We could also have a dictionary that maps a string to the function in python.  For instance:

```
validation_func_map = {
    "function1": my_function 
}

function_to_call = config['function_to_call']
validation_func_map[function_to_call](**kwargs)
```

Reply:
Great suggestion.  Added the function map as a dictionary and added a test to ensure correspondence between the config file and the map.